core/tests/mod.rs

```diff
@@ -0,0 +1,646 @@
+                    }
+                    if correct_count + faulty_peers.lock().unwrap().len() >= peer_ids.len() {
+                        success = true;
+                        break;
+                    }
+                }
+                sleep_ms(100);
+            }
+            assert!(success, "Not all sumeragi instances committed the block.");
+        }
+
+        {
+            let rng = &mut rand::rngs::StdRng::from_entropy();
+            let mut fp = faulty_peers.lock().unwrap();
+            fp.clear();
+            fp.extend(peer_ids.iter().cloned().choose_multiple(rng, fault_count));
+        }
+
+        let account_id: AccountId = "alice@wonderland".parse().unwrap();
+        let asset_definition_id: AssetDefinitionId = "rose#wonderland".parse().expect("Valid");
+        let mint_asset = MintBox::new(
+            1_u32.to_value(),
+            IdBox::AssetId(AssetId::new(
+                asset_definition_id.clone(),
+                account_id.clone(),
+            )),
+        );
+
+        let transaction =
+            TransactionBuilder::new(account_id.clone(), vec![mint_asset.into()], 100_000)
+                .sign(get_key_pair())
+                .unwrap();
+        let transaction = AcceptedTransaction::accept::<false>(
+            transaction,
+            &blank_config().wsv.transaction_limits,
+        )
+        .unwrap();
+
+        for peer_id in &peer_ids {
+            if !faulty_peers.lock().unwrap().contains(peer_id) {
+                let sumeragis = sumeragis.lock().unwrap();
+                let wsv = sumeragis
+                    .get(peer_id)
+                    .unwrap()
+                    .apply_wsv(WorldStateView::clone);
+                let queue = queues.get(peer_id).unwrap();
+                queue.push(transaction.clone().into(), &wsv).unwrap();
+            }
+        }
+    }
+}
+
+/*  This test does not actually verify that soft forking occured. It will trigger them only. Grep the log of this test for "Soft"
+to ensure soft forks are actually happening. If soft fork recovery is broken then this test will fail. */
+#[test]
+fn sumeragi_softfork() {
+    init_telemetry();
+
+    let should_softfork = Arc::new(Mutex::new(false));
+    let local_should_softfork = Arc::clone(&should_softfork);
+
+    let (_keypairs, peer_ids, _configs, queues, sumeragis) = set_up_network(
+        7,
+        Arc::new(move |packet, peer_id, sumeragis| {
+            if let Message::BlockCommitted(_) = packet.message {
+                let s = {
+                    let mut m = local_should_softfork.lock().unwrap();
+                    let s = *m;
+                    *m = false;
+                    s
+                };
+                if s {
+                    sleep_ms(20000);
+                }
+            }
+
+            let sumeragis = sumeragis.lock().unwrap();
+            sumeragis.get(peer_id).unwrap().incoming_message(packet);
+        }),
+    );
+
+    for block_height in 1..5 {
```

Comment:
I don't mind this not being a separate function, though you could extract a few things here as parameters. 

There's a lot of mutable state here though, which hurts readability. Loads of back-references. Can we try and do something about that?

Reply:
+1 for removing mutable variables